{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "from joblib import Parallel, delayed\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "sys.path.append('../../')\n",
    "\n",
    "\n",
    "from astroLuSt.monitoring import (\n",
    "    errorlogging as almoer,\n",
    "    errortesting as almoet,\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LogErrors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exception</th>\n",
       "      <th>prefix</th>\n",
       "      <th>suffix</th>\n",
       "      <th>file</th>\n",
       "      <th>line</th>\n",
       "      <th>problem_line</th>\n",
       "      <th>error_message</th>\n",
       "      <th>time</th>\n",
       "      <th>squeezed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf0</td>\n",
       "      <td>/tmp/ipykernel_201975/880230816.py</td>\n",
       "      <td>7</td>\n",
       "      <td>almoet.raise_concaterr(n)</td>\n",
       "      <td>ValueError: zero-dimensional arrays cannot be ...</td>\n",
       "      <td>2024-11-05 20:04:08.596059</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf0</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>12</td>\n",
       "      <td>return np.concatenate(n, 0)</td>\n",
       "      <td>ValueError: zero-dimensional arrays cannot be ...</td>\n",
       "      <td>2024-11-05 20:04:08.596059</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf2</td>\n",
       "      <td>/tmp/ipykernel_201975/880230816.py</td>\n",
       "      <td>9</td>\n",
       "      <td>almoet.raise_div0()</td>\n",
       "      <td>ZeroDivisionError: division by zero</td>\n",
       "      <td>2024-11-05 20:04:08.598997</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf2</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>9</td>\n",
       "      <td>return 1/0</td>\n",
       "      <td>ZeroDivisionError: division by zero</td>\n",
       "      <td>2024-11-05 20:04:08.598997</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf3</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>17</td>\n",
       "      <td>raise ValueError('custom error')</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.600480</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf3</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>21</td>\n",
       "      <td>raise ValueError('custom error in except')</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.600480</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf3</td>\n",
       "      <td>/tmp/ipykernel_201975/880230816.py</td>\n",
       "      <td>11</td>\n",
       "      <td>almoet.raise_nested_exception()</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.600480</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf3</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>23</td>\n",
       "      <td>raise ValueError('custom error in except in ex...</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.600480</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf4</td>\n",
       "      <td>/tmp/ipykernel_201975/880230816.py</td>\n",
       "      <td>9</td>\n",
       "      <td>almoet.raise_div0()</td>\n",
       "      <td>ZeroDivisionError: division by zero</td>\n",
       "      <td>2024-11-05 20:04:08.601886</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/t...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf4</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>9</td>\n",
       "      <td>return 1/0</td>\n",
       "      <td>ZeroDivisionError: division by zero</td>\n",
       "      <td>2024-11-05 20:04:08.601886</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf1</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>17</td>\n",
       "      <td>raise ValueError('custom error')</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.597741</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf1</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>21</td>\n",
       "      <td>raise ValueError('custom error in except')</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.597741</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf1</td>\n",
       "      <td>/tmp/ipykernel_201975/880230816.py</td>\n",
       "      <td>11</td>\n",
       "      <td>almoet.raise_nested_exception()</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.597741</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Traceback (most recent call last):\\n  File \"/h...</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Suf1</td>\n",
       "      <td>/home/lukas/github/astroLuSt/demos/monitoring/...</td>\n",
       "      <td>23</td>\n",
       "      <td>raise ValueError('custom error in except in ex...</td>\n",
       "      <td>ValueError: custom error in except in except</td>\n",
       "      <td>2024-11-05 20:04:08.597741</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           exception prefix suffix  \\\n",
       "0  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf0   \n",
       "0  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf0   \n",
       "1  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf2   \n",
       "1  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf2   \n",
       "2  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf3   \n",
       "2  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf3   \n",
       "2  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf3   \n",
       "2  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf3   \n",
       "3  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf4   \n",
       "3  Traceback (most recent call last):\\n  File \"/t...    Pre   Suf4   \n",
       "1  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf1   \n",
       "1  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf1   \n",
       "1  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf1   \n",
       "1  Traceback (most recent call last):\\n  File \"/h...    Pre   Suf1   \n",
       "\n",
       "                                                file line  \\\n",
       "0                 /tmp/ipykernel_201975/880230816.py    7   \n",
       "0  /home/lukas/github/astroLuSt/demos/monitoring/...   12   \n",
       "1                 /tmp/ipykernel_201975/880230816.py    9   \n",
       "1  /home/lukas/github/astroLuSt/demos/monitoring/...    9   \n",
       "2  /home/lukas/github/astroLuSt/demos/monitoring/...   17   \n",
       "2  /home/lukas/github/astroLuSt/demos/monitoring/...   21   \n",
       "2                 /tmp/ipykernel_201975/880230816.py   11   \n",
       "2  /home/lukas/github/astroLuSt/demos/monitoring/...   23   \n",
       "3                 /tmp/ipykernel_201975/880230816.py    9   \n",
       "3  /home/lukas/github/astroLuSt/demos/monitoring/...    9   \n",
       "1  /home/lukas/github/astroLuSt/demos/monitoring/...   17   \n",
       "1  /home/lukas/github/astroLuSt/demos/monitoring/...   21   \n",
       "1                 /tmp/ipykernel_201975/880230816.py   11   \n",
       "1  /home/lukas/github/astroLuSt/demos/monitoring/...   23   \n",
       "\n",
       "                                        problem_line  \\\n",
       "0                          almoet.raise_concaterr(n)   \n",
       "0                        return np.concatenate(n, 0)   \n",
       "1                                almoet.raise_div0()   \n",
       "1                                         return 1/0   \n",
       "2                   raise ValueError('custom error')   \n",
       "2         raise ValueError('custom error in except')   \n",
       "2                    almoet.raise_nested_exception()   \n",
       "2  raise ValueError('custom error in except in ex...   \n",
       "3                                almoet.raise_div0()   \n",
       "3                                         return 1/0   \n",
       "1                   raise ValueError('custom error')   \n",
       "1         raise ValueError('custom error in except')   \n",
       "1                    almoet.raise_nested_exception()   \n",
       "1  raise ValueError('custom error in except in ex...   \n",
       "\n",
       "                                       error_message  \\\n",
       "0  ValueError: zero-dimensional arrays cannot be ...   \n",
       "0  ValueError: zero-dimensional arrays cannot be ...   \n",
       "1                ZeroDivisionError: division by zero   \n",
       "1                ZeroDivisionError: division by zero   \n",
       "2       ValueError: custom error in except in except   \n",
       "2       ValueError: custom error in except in except   \n",
       "2       ValueError: custom error in except in except   \n",
       "2       ValueError: custom error in except in except   \n",
       "3                ZeroDivisionError: division by zero   \n",
       "3                ZeroDivisionError: division by zero   \n",
       "1       ValueError: custom error in except in except   \n",
       "1       ValueError: custom error in except in except   \n",
       "1       ValueError: custom error in except in except   \n",
       "1       ValueError: custom error in except in except   \n",
       "\n",
       "                        time  squeezed  \n",
       "0 2024-11-05 20:04:08.596059     False  \n",
       "0 2024-11-05 20:04:08.596059     False  \n",
       "1 2024-11-05 20:04:08.598997     False  \n",
       "1 2024-11-05 20:04:08.598997     False  \n",
       "2 2024-11-05 20:04:08.600480     False  \n",
       "2 2024-11-05 20:04:08.600480     False  \n",
       "2 2024-11-05 20:04:08.600480     False  \n",
       "2 2024-11-05 20:04:08.600480     False  \n",
       "3 2024-11-05 20:04:08.601886     False  \n",
       "3 2024-11-05 20:04:08.601886     False  \n",
       "1 2024-11-05 20:04:08.597741     False  \n",
       "1 2024-11-05 20:04:08.597741     False  \n",
       "1 2024-11-05 20:04:08.597741     False  \n",
       "1 2024-11-05 20:04:08.597741     False  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(almoer)\n",
    "\n",
    "def subfunc(idx):\n",
    "    n = np.random.choice(range(3), size=1)\n",
    "    try:\n",
    "        if n==0:\n",
    "            almoet.raise_concaterr(n)\n",
    "        elif n == 1:\n",
    "            almoet.raise_div0()\n",
    "        elif n == 2:\n",
    "            almoet.raise_nested_exception()\n",
    "    except Exception as e:\n",
    "        LE.print_exc(e, prefix='Pre', suffix=f'Suf{n}')\n",
    "        df_le_temp = LE.exc2df(e, prefix='Pre', suffix=f'Suf{idx}', store=True)\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "x = pd.DataFrame(\n",
    "    columns=['a', 'b'],\n",
    "#      data=[[1,2],[2,3]]\n",
    ")\n",
    "\n",
    "LE = almoer.LogErrors(verbose=0)\n",
    "\n",
    "\n",
    "_ = Parallel(n_jobs=5, backend='threading')(delayed(subfunc)(idx=idx) for idx in range(5))\n",
    "    \n",
    "\n",
    "# print(LE.df_errorlog['exception'].iloc[0])\n",
    "LE.df_errorlog\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "astroLuSt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
